### DISCUSSION TITLE
Confused about dokku logs missmatch output

### DISCUSSION THREAD

**Discussion by nhevia** [dokku/dokku#4756] [Q&A] [answered]
Hi. I need some help to understand how this log missmatch is happening. 

`dokku logs frontend` is showing `docker logs <container id>` and that is ok. The problem is that those logs are skipping my nginx `log_format` directives (in the nginx.conf main file). And that means I'm seeing nginx IP in the logs (172.0.0.1) instead of the actual forwarded one.

Now, nginx configuration for this app under `/home/dokku/frontend/nginx.conf` is `access_log  /var/log/nginx/frontend-access.log`. If I check those logs, they are not the same as `dokku logs frontend` (frontend-access.log is correctly taking my log_format directive inside nginx.conf main file).

The thing is, it seems like docker is generating its own log file, but according to [this](https://docs.docker.com/config/containers/logging/)  it shouldn't ?
> The official nginx image creates a symbolic link from /var/log/nginx/access.log to /dev/stdout, and creates another symbolic link from /var/log/nginx/error.log to /dev/stderr (...)

Is there anything else going under the hood? Can anyone please point me in the correct direction?

Thanks

**TL;DR:** `dokku logs frontend` shows a diff log than `/var/log/nginx/frontend-access.log`.

**Comment by Cellane** [comment, marked as answer]
I might be misunderstanding, but I think two different concepts are being mixed up here:

- `dokku logs <app>` outputs log generated by the application itself; this might include output to `stdout`, output from your logger library, and your own custom logging that you integrated in your app
- `dokku nginx:access-logs <app> [-t]` outputs/streams access logs from the nginx that Dokku installs and manages for you, in order to properly proxy and route incoming requests to appropriate containers

**Reply by nhevia** [reply]
Hi, thanks for the reply. 

How would you change the output from `dokku logs <app>` to match `nginx:access-logs <app>` then?

`dokku logs` format is similar to `access-logs`, but changing `etc/nginx/nginx.conf` (or any other conf file that is included into it, including dokku ones) is not affecting the log_format.

Adding this to `http` directive: 

```
 #log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
        #'$status $body_bytes_sent "$http_referer" '
        #'"$http_user_agent" "$http_x_forwarded_for"';
```

And changing `/home/dokku/frontend/nginx.conf` to `access_log  /var/log/nginx/frontend-access.log main;` makes no difference.

**Reply by Cellane** [reply]
> How would you change the output from dokku logs <app> to match nginx:access-logs <app> then?

I… would not, since both log streams have their own purpose 😅 But if really needed, then the only way I can think of is implementing your own request logging (independent of nginx) in your front-end application (if that’s possible) that would mirror the desired output format.

> `dokku logs` format is similar to `access-logs`

I just want to correct this statement, that’s not universally true. `dokku logs` streams `stdout` from your application container, it doesn’t have its own format. It simply shows what your application outputs. So if your application already outputs some sort of request log, you’ll need to be looking into how to adjust its format, rather than altering nginx configuration.

**Reply by josegonzalez** [reply]
If your app is running an internal nginx server (true for static deploys and some php ones), you'll need to change the format that server uses. That app nginx server is separate from the nginx that produces the access-logs/error-logs output. The former runs inside a container while the latter runs on the host, and thus they share no config.

**Reply by nhevia** [reply]
That is exactly the case, I'm using an static deploy. I'll check for the internal nginx config and change the log_format.
Thanks a lot you both for the help.
